### ISSUE TITLE
Tractography tool: Change colour handling UI

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#326]
In other tools, object colouring is handled by a combobox, so is visible at all times, But in the trqactography tool, you have to right-click on the entry in the list to change its colouring. Personally I find this inconsistent, even if it is a fration faster in some use cases.

Do people agree with making it consistent with other tools?

**Comment by thijsdhollander** [comment]
I remember this coming up a while ago as well, somewhere, as part of some.. issue.. somewhere.

Well, I'm definitely in favour of consistency; and even more so: not hiding things in non-obvious context menus. :+1:

**Comment by draffelt** [comment]
Agreed

**Comment by jdtournier** [comment]
Guess I'm outvoted on that one... And I can't exactly argue that the context menu is the most intuitive. It's almost impossible to get one on MacOSX anyway... Happy to go along with that.

**Comment by Lestropie** [comment]
See branch `mrview_tractography_overlay_colours`.

1. For the tractography tool, the right-click context menu remains unchanged, but there's also a colour control in the main tool UI. The colour button is only enabled if the tractogram is currently set to use a fixed colour.

2. For the overlay tool, I cheated a little: right-clicking on an entry in the file list simply selects that entry and then opens the colour map button's context menu.

3. I'm not sure if this should extend to other tools:
  - The ROI tool only has the colour button and no other controls, so if there were a right-click context menu, it would be limited to 'Set colour' / 'Random colour'. Still worth it?
  - The ODF display can only show one file at a time, so you're either colouring by direction or by a fixed colour (don't see any value for a random colour here); no need to go through input files changing one at a time.
  - The vector plot tool has both a 'Colour by (value / direction / fixed)' control, and a colour map button. So having a right-click context menu isn't quite as clean as for the overlay tool. And the colour controls are immediately below the file list. But it could conceivably be done.

4. In discussion with @draffelt , will try moving the (track) scalar file controls for the tractography tool into the main tool UI rather than acting as a dock, and have the whole thing disappear whenever it's not needed.

**Comment by Lestropie** [comment]
* Point 4 done.

* In addition, the controls for colouring by scalar file and thresholding by scalar file are much more independent; this is an expansion of the contents of the `scalar_thresh_hotfix` that was put together for the sake of an article figure but not fully fleshed out.

* I've re-arranged the UI elements in the tractography tool, so that per-tractogram settings are at the top and global tool settings are at the bottom.

**Comment by draffelt** [comment]
@Lestropie just had a play with this. It all looks good. However one issue I had is that I can no longer threshold by track scalar file, then switch back to colour by direction.  Previously this was available as an extra colour option in the colour map button menu. If I try and select colour as direction, then threshold by file, the current version returns this error:
_Can only threshold based on scalar file used for streamline colouring if that colour mode is active_

**Comment by Lestropie** [comment]
This _should_ work, it's just handled differently to before.

Previously, you would select 'colour by scalar file', then activate the thresholding by scalar file, then choose 'scalarfile by direction' in the colourmap menu, which is kind of hacky. Here, the thresholding by scalar file can be done _either_ using the same scalar file as that used for colouring, _or_ using a scalar file completely independent to the track colouring.

So if you want to threshold by scalar file but leave the colouring based on track direction, _don't_ select colourby scalar file; just leave the track colouring as 'by direction' and open a scalar file specifically for thresholding.

**Comment by draffelt** [comment]
I did try that however nothing happened when I moved the threshold adjust button. I've just tried again and it does work, but only if you set the threshold manually. Realising there must be an issue with the button change rate, I had a look into the code and saw your FIXME:  
```
// FIXME Required scaling rate for threshold must be independent of colour
```
I'm happy to fix it #AfterISMRM.

**Comment by Lestropie** [comment]
This should be working now, but I'm finding I can cause a crash when going from colouring by scalar file to colouring by streamline endpoints, and it's crashing too hard for me to catch.

**Comment by Lestropie** [comment]
Closed by #611.
